File: CSC330/poker/Cpp/player.cpp
#include "player.h"
#include <algorithm>
#include <iostream>

using namespace std;
Player::Player() {

}
std::vector<Card> Player::getHand() const{
	return hand;
}
void Player::dealCard(const Card& card) {
    if (hand.size() < 5) {
        hand.push_back(card);
    }
}

void Player::sortByRank() {
    sort(hand.begin(), hand.end(), [](const Card& a, const Card& b) {
        return a.getRankValue() > b.getRankValue();
    });
}

void Player::sortBySuit() {
    sort(hand.begin(), hand.end(), [](const Card& a, const Card& b) {
        return a.getSuitValue() > b.getSuitValue();
    });
}

bool Player::isRoyal() const{
    return isFlush() && isStraight() && hand.back().getRank() == 'T';
}

bool Player::isStraightFlush() const {
    return isFlush() && isStraight();
}

bool Player::isFour() const {
    return (hand[0].getRank() == hand[1].getRank() &&
            hand[1].getRank() == hand[2].getRank() &&
            hand[2].getRank() == hand[3].getRank()) ||
           (hand[1].getRank() == hand[2].getRank() &&
            hand[2].getRank() == hand[3].getRank() &&
            hand[3].getRank() == hand[4].getRank());
}

bool Player::isFull() const{
    return (hand[0].getRank() == hand[1].getRank() &&
            hand[1].getRank() == hand[2].getRank() &&
            hand[3].getRank() == hand[4].getRank()) ||
           (hand[2].getRank() == hand[3].getRank() &&
            hand[3].getRank() == hand[4].getRank() &&
            hand[0].getRank() == hand[1].getRank());
}

bool Player::isFlush() const{
    return hand[0].getSuit() == hand[1].getSuit() &&
           hand[1].getSuit() == hand[2].getSuit() &&
           hand[2].getSuit() == hand[3].getSuit() &&
           hand[3].getSuit() == hand[4].getSuit();
}

bool Player::isStraight() const{
    for (int i = 0; i < 4; ++i) {
        if (hand[i].getRankValue() - 1 != hand[i + 1].getRankValue()) {
            return false;
        }
    }
    return true;
}

bool Player::isThree() const{
    return (hand[0].getRank() == hand[1].getRank() &&
            hand[1].getRank() == hand[2].getRank()) ||
           (hand[1].getRank() == hand[2].getRank() &&
            hand[2].getRank() == hand[3].getRank()) ||
           (hand[2].getRank() == hand[3].getRank() &&
            hand[3].getRank() == hand[4].getRank());
}

bool Player::isTwo() const{
    return (hand[0].getRank() == hand[1].getRank() &&
            hand[2].getRank() == hand[3].getRank()) ||
           (hand[1].getRank() == hand[2].getRank() &&
            hand[3].getRank() == hand[4].getRank());
}

bool Player::isPair() const{
    return hand[0].getRank() == hand[1].getRank() ||
           hand[1].getRank() == hand[2].getRank() ||
           hand[2].getRank() == hand[3].getRank() ||
           hand[3].getRank() == hand[4].getRank();
}

int Player::handRank() const{
    if (isRoyal()) return 10;
    if (isStraightFlush()) return 9;
    if (isFour()) return 8;
    if (isFull()) return 7;
    if (isFlush()) return 6;
    if (isStraight()) return 5;
    if (isThree()) return 4;
    if (isTwo()) return 3;
    if (isPair()) return 2;
    return 1;
}

void Player::printHand() {
    for (const Card& card : hand) {
        cout << card.toString() << " ";
    }
    cout << endl;
}

File: CSC330/poker/C#/player.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Player
{
    private List<Card> hand;

    public Player()
    {
        hand = new List<Card>();
    }

    public List<Card> GetHand()
    {
        return hand;
    }

    public void DealCard(Card card)
    {
        if (hand.Count < 5)
        {
            hand.Add(card);
        }
    }

    public void SortByRank()
    {
        hand = hand.OrderByDescending(c => c.GetRankValue()).ToList();
    }

    public void SortBySuit()
    {
        hand = hand.OrderByDescending(c => c.GetSuitValue()).ToList();
    }

    public bool IsRoyal()
    {
        return IsFlush() && IsStraight() && hand[^1].GetRank() == 'T';
    }

    public bool IsStraightFlush()
    {
        return IsFlush() && IsStraight();
    }

    public bool IsFour()
    {
        return (hand[0].GetRank() == hand[1].GetRank() &&
                hand[1].GetRank() == hand[2].GetRank() &&
                hand[2].GetRank() == hand[3].GetRank()) ||
               (hand[1].GetRank() == hand[2].GetRank() &&
                hand[2].GetRank() == hand[3].GetRank() &&
                hand[3].GetRank() == hand[4].GetRank());
    }

    public bool IsFull()
    {
        return (hand[0].GetRank() == hand[1].GetRank() &&
                hand[1].GetRank() == hand[2].GetRank() &&
                hand[3].GetRank() == hand[4].GetRank()) ||
               (hand[2].GetRank() == hand[3].GetRank() &&
                hand[3].GetRank() == hand[4].GetRank() &&
                hand[0].GetRank() == hand[1].GetRank());
    }

    public bool IsFlush()
    {
        return hand.All(c => c.GetSuit() == hand[0].GetSuit());
    }

    public bool IsStraight()
    {
        for (int i = 0; i < 4; i++)
        {
            if (hand[i].GetRankValue() - 1 != hand[i + 1].GetRankValue())
            {
                return false;
            }
        }
        return true;
    }

    public bool IsThree()
    {
        return (hand[0].GetRank() == hand[1].GetRank() && hand[1].GetRank() == hand[2].GetRank()) ||
               (hand[1].GetRank() == hand[2].GetRank() && hand[2].GetRank() == hand[3].GetRank()) ||
               (hand[2].GetRank() == hand[3].GetRank() && hand[3].GetRank() == hand[4].GetRank());
    }

    public bool IsTwo()
    {
        return (hand[0].GetRank() == hand[1].GetRank() && hand[2].GetRank() == hand[3].GetRank()) ||
               (hand[1].GetRank() == hand[2].GetRank() && hand[3].GetRank() == hand[4].GetRank());
    }

    public bool IsPair()
    {
        return hand[0].GetRank() == hand[1].GetRank() ||
               hand[1].GetRank() == hand[2].GetRank() ||
               hand[2].GetRank() == hand[3].GetRank() ||
               hand[3].GetRank() == hand[4].GetRank();
    }

    public int HandRank()
    {
        if (IsRoyal()) return 10;
        if (IsStraightFlush()) return 9;
        if (IsFour()) return 8;
        if (IsFull()) return 7;
        if (IsFlush()) return 6;
        if (IsStraight()) return 5;
        if (IsThree()) return 4;
        if (IsTwo()) return 3;
        if (IsPair()) return 2;
        return 1;
    }

    public void PrintHand()
    {
        foreach (Card card in hand)
        {
            Console.Write(card.ToString() + " ");
        }
        
    }
}

